### PR TITLE
asn1_viewer

### PR DESCRIPTION
**Pull request by SergZen**

### CODE REVIEW

**Review by Jujumba — APPROVED** [review]

File: src/input/mod.rs
Comment (ikripaka) [inline src/input/mod.rs:1]: Consider moving logic from mod.rs file into inner one to leave it clean from other code. It'll make your code to look more tidy. Link to article about this: [link](https://kerkour.com/rust-patterns-dont-put-code-in-lib-mod-files).
For example:
* move declaration of error into -> `error.rs` in the root of the project
* move declaration of cli into -> `cli.rs` in the root of the project
* and other functions into `utils.rs`

File: src/input/mod.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+use crate::input::base64::{base64_decode, is_valid_base64};
+use crate::input::hex::{hex_decode, is_valid_hex};
+
+pub mod base64;
+pub mod hex;
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Cli {
+    #[arg(short = 'f', long, value_name = "FILE")]
```
Comment (ikripaka) [inline src/input/mod.rs]: add about attribute to struct derive
https://docs.rs/clap/latest/clap/_derive/index.html#command-attributes

File: src/input/mod.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    file: Option<PathBuf>,
+
+    #[arg(short = 'a', long)]
```
Comment (ikripaka) [inline src/input/mod.rs]: add about attribute to struct derive
https://docs.rs/clap/latest/clap/_derive/index.html#command-attributes

**Review by ikripaka — APPROVED** [review]
As far as looked on this project and tried to run it, it look great! I like your idea of traversing ASN1 tree!

On the other hand, I have some problems with building and running your project on linux. When I ran it, the terminal window appeared, but any button that I pressed couldn't be captured by the terminal. As I understand your app uses a specific terminal that works correctly only in Windows Cmd terminal. (or I have the bad version of Ubuntu)
Moreover, your project lacks documentation and comments to your code. When I ran a command `asn1_viewer -asn1`, there were no helpful message of what this flag is doing and what the input for this command is.

So, to sum up, you have to consider doing such things to make your more convenient to use:
* add docs on how to use your app, especially in cli interface (in derived clap structure) and comment on your implemented functions using doc comments;
(furthermore, you can export your REAMDE.md file into the project feed for using clear view in the `cargo doc` resulting web page)
* fix terminal button capture on unix systems or explicitly write that it will work only in Windows Cmd terminal.

File: src/input/mod.rs
Comment (SergZen) [reply src/input/mod.rs:1]: I added  comments every args for cli. Its will be show in --help.
I fixed the problem with terminal buttons capture

File: src/input/mod.rs (hunk 2 above)
Comment (SergZen) [reply src/input/mod.rs]: I added comments for every args. Attribute "about" is working only with command derive

File: src/input/mod.rs
Comment (SergZen) [reply src/input/mod.rs:1]: Good idea about refactoring mod input. I made changes following your idea

**Review by TheBestTvarynka — APPROVED** [review]

**Review by ikripaka — APPROVED** [review]

**Review by alex4ip — APPROVED** [review]
